javalin/src/main/java/io/javalin/http/staticfiles/PrecompressingResourceHandler.kt

```diff
@@ -0,0 +1,109 @@
+
+    companion object {
+        val excludedMimeTypes = setOf(
+                "image/",
+                "audio/",
+                "video/",
+                "application/compress",
+                "application/zip",
+                "application/gzip",
+                "application/bzip2",
+                "application/brotli",
+                "application/x-xz",
+                "application/x-rar-compressed")
+    }
+
+    fun handle(resource: Resource, req: HttpServletRequest, res: HttpServletResponse): Boolean {
+        if (resource.exists() && !resource.isDirectory) {
+            val target = req.getAttribute("jetty-target") as String
+            var acceptCompressType = CompressType.getByAcceptEncoding(req.getHeader(Header.ACCEPT_ENCODING) ?: "")
+            val contentType = MimeTypes.getDefaultMimeByExtension(resource.file.name)//get content type by file extension
+            if (excludedMimeType(contentType))
+                acceptCompressType = CompressType.NONE
+            val resultByteArray = getStaticResourceByteArray(resource, target, acceptCompressType) ?: return false
+            res.setContentLength(resultByteArray.size)
+            res.setHeader(Header.CONTENT_TYPE, contentType)
+
+            if (acceptCompressType != CompressType.NONE)
+                res.setHeader(Header.CONTENT_ENCODING, acceptCompressType.typeName)
+
+            val weakETag = resource.weakETag //jetty resource use weakETag too
+            req.getHeader(Header.IF_NONE_MATCH)?.let { etag ->
+                if (etag == weakETag) {
+                    res.status = 304
+                    return true
+                }
+            }
+            res.setHeader(Header.ETAG, weakETag)
+            resultByteArray.inputStream().copyTo(res.outputStream)
+            res.outputStream.close()
+            return true
+        }
+        return false
+    }
+
+    private fun getStaticResourceByteArray(resource: Resource, target: String, type: CompressType): ByteArray? {
+        if (resource.length() > resourceMaxSize) {
+            Javalin.log?.warn("PrecompressingResourceHandler can't handle " +
+                    "the resource of path : ${resource.file.path} with size : ${resource.length()} byte, " +
+                    "because it is larger than resourceMaxSize : $resourceMaxSize byte")
+            return null
+        }
+        return compressedFiles.computeIfAbsent(target + type.extension) { getCompressedByteArray(resource, type) }
+    }
+
+    private fun getCompressedByteArray(resource: Resource, type: CompressType): ByteArray {
+        val fileInput = resource.file.inputStream()
+        val byteArrayOutputStream = ByteArrayOutputStream()
+        val outputStream: OutputStream = when {
+            type == CompressType.GZIP -> {
+                LeveledGzipStream(byteArrayOutputStream, 9) // use max-level compression
+            }
+            BrotliLoader.isBrotliAvailable() && type == CompressType.BR -> {
```

Comment:
You can't use this function without already needing the dependency it's from. You need to use the `OptionalDependency` enum. Maybe add something like this to the `Util` class? 


```kotlin
fun dependencyIsPresent(dependency: OptionalDependency) = try {
    ensureDependencyPresent(dependency)
    true
} catch (e: Exception) {
    false
}
```

Reply:
Got it.
I think I will take your example code to add.